pkg/syncer/synctarget/controller.go

```diff
@@ -0,0 +1,257 @@
+		FilterFunc: func(obj interface{}) bool {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err != nil {
+				return false
+			}
+			_, name, err := cache.SplitMetaNamespaceKey(key)
+			if err != nil {
+				return false
+			}
+			return name == syncTargetName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.enqueueSyncTarget(obj, logger) },
+			UpdateFunc: func(old, obj interface{}) { c.enqueueSyncTarget(obj, logger) },
+			DeleteFunc: func(obj interface{}) { c.enqueueSyncTarget(obj, logger) },
+		},
+	})
+
+	return c, nil
+}
+
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	syncTargetUID    types.UID
+	syncTargetLister workloadv1alpha1listers.SyncTargetLister
+	syncTargetClient workloadv1alpha1typed.SyncTargetInterface
+
+	shardManager *shardManager
+	gvrSource    *syncTargetGVRSource
+}
+
+func (c *controller) enqueueSyncTarget(obj interface{}, logger logr.Logger) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logging.WithQueueKey(logger, key).V(2).Info("queueing SyncTarget")
+
+	c.queue.Add(key)
+}
+
+// Start starts the controller worker.
+func (c *controller) Start(ctx context.Context) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), controllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("failed to sync %q: %w", key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "failed to split key, dropping")
+		return nil
+	}
+
+	syncTarget, err := c.syncTargetLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		c.shardManager.updateShards(ctx, nil)
+		c.gvrSource.updateGVRs(ctx, nil)
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if syncTarget.GetUID() != c.syncTargetUID {
+		return nil
+	}
```

Comment:
it is strange to stop controller on not-found, but not here

Reply:
Fixed. I also stop the controller in this case now.